File: alphahunter-ai/backend/exit_rules.py
"""Exit rules — when to take the money.

The product has always said Buy and never said Sell. That is not a small gap:
the portfolio backtest in this repo found that holding the picks for 10 trading
days returned +21% while holding them indefinitely (the paper portfolio) came
out roughly flat and behind SPY. Same picks. The entire difference is the exit.

So every recommendation now carries a plan with four ways out, checked in the
order a real trader checks them:

  1. STOP   — the thesis is wrong. Cut it. Checked first, always: if a gap
              takes price through both the stop and the target, the loss is
              what actually happened to you.
  2. TARGET — the move played out. Book it.
  3. TIME   — the edge measured here is a multi-week bounce, not a multi-year
              hold. If nothing has happened by the horizon, the reason for
              owning it has expired.
  4. TRAIL  — once a position is up enough to matter, protect it. A winner
              that round-trips to breakeven is the most demoralising outcome
              in the book and the easiest to prevent.

Pure functions: prices and a plan in, a decision out. No network, no state.
"""
from __future__ import annotations

from dataclasses import dataclass, asdict

# Defaults come from this repo's own measurements, not from convention.
DEFAULT_HORIZON_DAYS = 10       # the holding period the backtest actually paid
DEFAULT_TARGET_PCT = 12.0       # take-profit, ~2x the typical 10-day move
DEFAULT_STOP_PCT = -7.0         # cut, sized so target:stop is better than 1.5:1
TRAIL_ARMS_AT_PCT = 8.0         # only start trailing once there is a gain worth keeping
TRAIL_GIVEBACK_PCT = 5.0        # ...then give back at most this much from the peak


@dataclass
class ExitPlan:
    entry: float
    target: float
    stop: float
    horizon_days: int
    target_pct: float
    stop_pct: float
    trail_arms_at_pct: float
    trail_giveback_pct: float

    def to_dict(self) -> dict:
        return asdict(self)


def build_plan(
    entry: float,
    *,
    atr: float | None = None,
    horizon_days: int = DEFAULT_HORIZON_DAYS,
    target_pct: float = DEFAULT_TARGET_PCT,
    stop_pct: float = DEFAULT_STOP_PCT,
) -> ExitPlan:
    """Target and stop for one position.

    When ATR is known the levels widen or tighten to the stock's own volatility
    — a fixed 7% stop is noise on a name that moves 4% a day and a straitjacket
    on one that moves 0.5% — but they are clamped so a wild reading cannot
    produce an absurd plan.
    """
    if entry <= 0:
        raise ValueError("entry must be positive")

    tp, sp = target_pct, stop_pct
    if atr and atr > 0:
        atr_pct = atr / entry * 100
        # 2 ATR to the target, 1.2 ATR to the stop, then clamped to sane bounds.
        tp = max(6.0, min(30.0, atr_pct * 2.0))
        sp = -max(3.0, min(15.0, atr_pct * 1.2))

    return ExitPlan(
        entry=round(entry, 2),
        target=round(entry * (1 + tp / 100), 2),
        stop=round(entry * (1 + sp / 100), 2),
        horizon_days=horizon_days,
        target_pct=round(tp, 2),
        stop_pct=round(sp, 2),
        trail_arms_at_pct=TRAIL_ARMS_AT_PCT,
        trail_giveback_pct=TRAIL_GIVEBACK_PCT,
    )


def check_exit(
    plan: ExitPlan,
    price: float,
    *,
    days_held: int = 0,
    peak_price: float | None = None,
) -> dict:
    """Should this position be closed now, and why?

    ``peak_price`` is the highest price seen since entry; without it the trail
    cannot be evaluated and is simply skipped rather than guessed at.
    """
    gain_pct = (price / plan.entry - 1) * 100
    result = {
        "action": "hold",
        "reason": "",
        "gain_%": round(gain_pct, 2),
        "days_held": days_held,
    }

    # 1. Stop first — if both levels were breached, the loss is the real event.
    if price <= plan.stop:
        result.update(action="sell", reason=(
            f"stop hit at ${plan.stop:.2f} ({plan.stop_pct:.1f}%). The reason for "
            f"owning it is gone; take the small loss."))
        return result

    # 2. Target.
    if price >= plan.target:
        result.update(action="take_profit", reason=(
            f"target hit at ${plan.target:.2f} ({plan.target_pct:+.1f}%). "
            f"Book it — this is the move the setup was predicting."))
        return result

    # 3. Trailing stop, once there is a real gain to protect.
    if peak_price is not None and peak_price > plan.entry:
        peak_gain = (peak_price / plan.entry - 1) * 100
        giveback = (peak_price - price) / peak_price * 100
        if peak_gain >= plan.trail_arms_at_pct and giveback >= plan.trail_giveback_pct:
            result.update(action="take_profit", reason=(
                f"was up {peak_gain:.1f}% and has given back {giveback:.1f}% from "
                f"the high. Protect the gain rather than watch it round-trip."))
            return result

    # 4. Time. The measured edge is a multi-week move; past that the thesis
    #    has simply not worked, whatever the price is doing.
    if days_held >= plan.horizon_days:
        result.update(action="close_stale", reason=(
            f"{days_held} trading days held with no target or stop hit "
            f"({gain_pct:+.1f}%). The setup's window has passed — free the capital."))
        return result

    left = plan.horizon_days - days_held
    result["reason"] = (
        f"{gain_pct:+.1f}% · target ${plan.target:.2f} / stop ${plan.stop:.2f} · "
        f"{left} day{'s' if left != 1 else ''} left in the window")
    return result

File: alphahunter-ai/frontend/src/lib/exitRules.ts
// Exit rules, client side.
//
// SOURCE OF TRUTH: backend/exit_rules.py. This is a deliberate port, not a
// second design — the constants and the order of checks must match it. It
// exists because the static deploy has no Python at request time, and asking a
// serverless function per holding would make the page slow for no benefit; the
// arithmetic is trivial.
//
// If you change a threshold, change it in both places.

export const DEFAULT_HORIZON_DAYS = 10;
export const DEFAULT_TARGET_PCT = 12.0;
export const DEFAULT_STOP_PCT = -7.0;
export const TRAIL_ARMS_AT_PCT = 8.0;
export const TRAIL_GIVEBACK_PCT = 5.0;

export type ExitPlan = {
  entry: number; target: number; stop: number;
  horizonDays: number; targetPct: number; stopPct: number;
};

export type ExitAction = "hold" | "take_profit" | "sell" | "close_stale";
export type ExitCheck = {
  action: ExitAction; reason: string; gainPct: number; daysHeld: number;
};

export function buildPlan(
  entry: number,
  opts: { atr?: number | null; horizonDays?: number } = {},
): ExitPlan {
  const horizonDays = opts.horizonDays ?? DEFAULT_HORIZON_DAYS;
  let targetPct = DEFAULT_TARGET_PCT;
  let stopPct = DEFAULT_STOP_PCT;

  if (opts.atr && opts.atr > 0 && entry > 0) {
    // Scale to the HOLDING PERIOD, not a single day: volatility grows with the
    // square root of time, so a 2%-a-day stock has a ~6.5% range over 10
    // sessions. Clamped so a bad ATR reading can't produce an absurd plan.
    const atrPct = (opts.atr / entry) * 100;
    const horizonMove = atrPct * Math.sqrt(horizonDays);
    targetPct = Math.max(5, Math.min(30, horizonMove));
    stopPct = -Math.max(3, Math.min(15, horizonMove * 0.6));   // ~1.67:1
  }

  return {
    entry: round2(entry),
    target: round2(entry * (1 + targetPct / 100)),
    stop: round2(entry * (1 + stopPct / 100)),
    horizonDays,
    targetPct: round2(targetPct),
    stopPct: round2(stopPct),
  };
}

export function checkExit(
  plan: ExitPlan,
  price: number,
  opts: { daysHeld?: number; peakPrice?: number | null } = {},
): ExitCheck {
  const daysHeld = opts.daysHeld ?? 0;
  const gainPct = (price / plan.entry - 1) * 100;
  const base = { gainPct: round2(gainPct), daysHeld };

  // 1. Stop first. If a gap took price through both levels, the loss is what
  //    actually happened to you.
  if (price <= plan.stop) {
    return { ...base, action: "sell", reason:
      `Stop hit at $${plan.stop.toFixed(2)} (${plan.stopPct.toFixed(1)}%). The reason for owning it is gone — take the small loss.` };
  }

  // 2. Target.
  if (price >= plan.target) {
    return { ...base, action: "take_profit", reason:
      `Target hit at $${plan.target.toFixed(2)} (+${plan.targetPct.toFixed(1)}%). Book it — this is the move the setup was predicting.` };
  }

  // 3. Trailing stop, once there is a gain worth protecting.
  if (opts.peakPrice != null && opts.peakPrice > plan.entry) {
    const peakGain = (opts.peakPrice / plan.entry - 1) * 100;
    const giveback = ((opts.peakPrice - price) / opts.peakPrice) * 100;
    if (peakGain >= TRAIL_ARMS_AT_PCT && giveback >= TRAIL_GIVEBACK_PCT) {
      return { ...base, action: "take_profit", reason:
        `Was up ${peakGain.toFixed(1)}% and has given back ${giveback.toFixed(1)}% from the high. Protect the gain rather than watch it round-trip.` };
    }
  }

  // 4. Time. The measured edge is a multi-week move; past that the thesis has
  //    not worked, whatever the price is doing.
  if (daysHeld >= plan.horizonDays) {
    return { ...base, action: "close_stale", reason:
      `${daysHeld} trading days held with no target or stop hit (${gainPct >= 0 ? "+" : ""}${gainPct.toFixed(1)}%). The setup's window has passed — free the capital.` };
  }

  const left = plan.horizonDays - daysHeld;
  return { ...base, action: "hold", reason:
    `${gainPct >= 0 ? "+" : ""}${gainPct.toFixed(1)}% · target $${plan.target.toFixed(2)} / stop $${plan.stop.toFixed(2)} · ${left} day${left === 1 ? "" : "s"} left in the window` };
}

/** Trading days between two ISO dates (weekends excluded, holidays ignored). */
export function tradingDaysBetween(fromISO: string, toDate = new Date()): number {
  const from = new Date(fromISO + "T00:00:00");
  if (Number.isNaN(from.getTime())) return 0;
  let days = 0;
  const cur = new Date(from);
  while (cur < toDate) {
    cur.setDate(cur.getDate() + 1);
    const d = cur.getDay();
    if (d !== 0 && d !== 6) days++;
  }
  return days;
}

const round2 = (v: number) => Math.round(v * 100) / 100;

export const ACTION_LABEL: Record<ExitAction, string> = {
  take_profit: "TAKE PROFIT",
  sell: "SELL — STOP",
  close_stale: "CLOSE — STALE",
  hold: "HOLD",
};
